src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -635,29 +613,33 @@ internal void FinishConnectByNameSyncFailure(Exception exception, int bytesTrans
 
         internal void FinishOperationAsyncFailure(SocketError socketError, int bytesTransferred, SocketFlags flags)
         {
+            ExecutionContext context = _context; // store context before it's cleared as part of finishing the operation
+
             FinishOperationSyncFailure(socketError, bytesTransferred, flags);
 
-            if (_context == null)
+            if (context == null)
```

Comment:
Since this pattern is repeated a couple times, couldn't we encapsulate it into something like "InvokeCompletion"?

And couldn't we move the clearing of the execution context there?

Reply:
Ok.